openlibrary/plugins/upstream/mybooks.py

```diff
@@ -587,3 +613,83 @@ def get_counts(cls, username):
             'observations': Observations.count_distinct_observations(username),
         }
+
+
+def get_loan_history_data(page: int, mb: "MyBooksTemplate") -> dict[str, str | int]:
+    """
+    Retrieve IA loan history data for page `page` of the patron's history.
+
+    This will use a patron's S3 keys to query the IA loan history API,
+    get the IA IDs, get the OLIDs if available, and and then convert this
+    into editions and IA-only items for display in the loan history.
+
+    This returns both editions and IA-only items because the loan history API
+    includes items that are not in Open Library, and displaying only IA
+    items creates pagination and navigation issues. For further discussion,
+    see https://github.com/internetarchive/openlibrary/pull/8375.
+    """
+    if not (account := OpenLibraryAccount.get(username=mb.username)):
+        raise render.notfound(
+            "Account for not found for %s" % mb.username, create=False
+        )
+    s3_keys = web.ctx.site.store.get(account._key).get('s3_keys')
+    limit = RESULTS_PER_PAGE
+    offset = page * limit - limit
+    loan_history = s3_loan_api(
+        s3_keys=s3_keys,
+        action='user_borrow_history',
+        limit=limit + 1,
+        offset=offset,
+        newest=True,
+    ).json()['history']['items']
+
+    # We request limit+1 to see if there is another page of history to display,
+    # and then pop the +1 off if it's present.
+    show_next = len(loan_history) == limit + 1
+    if show_next:
+        loan_history.pop()
+
+    ocaids = [loan_record['identifier'] for loan_record in loan_history]
+    availability = get_availability_of_ocaids(ocaids)
+    loan_history_map = {
+        loan_record['identifier']: loan_record for loan_record in loan_history
+    }
+    editions = web.ctx.site.get_many(
+        [
+            '/books/%s' % availability[ocaid].get('openlibrary_edition')
+            for ocaid in availability
+            if availability[ocaid].get('openlibrary_edition')
+        ]
+    )
+
+    # Create 'placeholder' editions for items in the Internet Archive loan
+    # history, but absent from Open Library.
+    ia_only_loans = [
+        loan for loan in availability.values() if not loan.get('openlibrary_edition')
+    ]
+
+    # Attach availability and loan to both editions and ia-only items.
+    for ed in editions:
+        if ed.ocaid in ocaids:
+            ed.availability = availability.get(ed.ocaid)
+            ed.loan = loan_history_map[ed.ocaid]
+            ed.last_loan_date = ed.loan.get('updatedate')
+
+    for ia_only in ia_only_loans:
+        loan = loan_history_map[ia_only['identifier']]
+        ia_only['last_loan_date'] = loan.get('updatedate', '')
+        ia_only['ia_only'] = True  # Determines which macro to load.
+
+    editions_and_ia_loans = editions + ia_only_loans
+    editions_and_ia_loans.sort(
+        key=lambda item: item.get('last_loan_date', ''), reverse=True
+    )
```

Comment:
Let's check against the `add_availability` function to see if this may diverge (and if there's a chance to DRY this code so we don't have to re-implement the add_availability logic)